glusterd2/brickmux/demultiplex.go

```diff
@@ -0,0 +1,58 @@
+package brickmux
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/gluster/glusterd2/glusterd2/brick"
+	"github.com/gluster/glusterd2/glusterd2/daemon"
+	"github.com/gluster/glusterd2/glusterd2/pmap"
+)
+
+// IsLastBrickInProc returns true if the brick specified is the last brick
+// present in the brick process.
+func IsLastBrickInProc(b brick.Brickinfo) bool {
+
+	port, err := pmap.RegistrySearch(b.Path)
+	if err != nil {
+		return false
+	}
+
+	return len(pmap.GetBricksOnPort(port)) == 1
+}
+
+// Demultiplex sends a detach request to the brick process which the
+// specified brick is multiplexed onto.
+func Demultiplex(b brick.Brickinfo) error {
+
+	brickDaemon, err := brick.NewGlusterfsd(b)
+	if err != nil {
+		return err
+	}
+
+	client, err := daemon.GetRPCClient(brickDaemon)
+	if err != nil {
+		return err
+	}
+
+	req := &brick.GfBrickOpReq{
+		Name: b.Path,
+		Op:   int(brick.OpBrickTerminate),
+	}
+
+	var rsp brick.GfBrickOpRsp
+	if err := client.Call("Brick.OpBrickTerminate", req, &rsp); err != nil {
+		return err
+	}
+
+	if rsp.OpRet != 0 {
+		return fmt.Errorf("detach brick RPC request failed; OpRet = %d", rsp.OpRet)
+	}
+
+	time.Sleep(time.Second * 1)
```

Comment:
What's the reason of adding this sleep if the rpc call and callback are synchronous? Also shouldn't we be marking brick status as stopped in case op.ret is 0?

Reply:
@atinmu So, the reason for adding this is to see if my doubts on client.Call being asynchronous was true. Now, I have seen a better result with adding sleep before we delete the socket file. But if @kshlm says that its synchronous then I don't feel there is any need for this sleep. But at the same time after looking at the e2e test results I am thinking that the EOF error could occur due to deleting the socket file even before the rpc result has returned.

Reply:
Another Question that I had is - Whether the detach request returns only after the brick has been detached and a signout request has been sent. Or  glusterfsd detects its a valid request, starts the sign out process and returns like what happens in heal(glustershd gets the heal RPC req, starts the heal process and returns with an output that  heal has started and then runs heal as an independent activity.)

Reply:
good question. detach request returns as long as rpc request succeeds and the brick graph removal from glusterfsd stack happens asynchronously.  Comparing the code of glusterd_volume_stop_glusterfs () from GD1, it looks like we mark this brick's rpc connection as NULL post sending the detach request. Is this what we're missing here?

Reply:
I am not entirely sure about this but what we have to make sure is that all the operations related to marking a brick as stopped(such as a signout request) should happen before we delete the socketfile and pidfile. That is why I added a sleep so that we can give some  time for all these processes before we delete socketfile. Although adding sleep isn't a very good preventive measure.